corehq/motech/repeaters/models.py

```diff
@@ -1395,6 +1412,14 @@ def later_might_be_better(resp):
         log_repeater_timeout_in_datadog(repeat_record.domain)
         message = str(RequestConnectionError(err))
         repeat_record.add_server_failure_attempt(message)
+    except RequestThrottled:
```

Comment:
@Charl1996 based on the discussion we had: `connection_settings.request_is_throttled()` basically looks at when last this **connection** was used to make a request. This is irrespective of the Repeater that was used with this connection. For example, say I set the throttle interval to 100ms, then I expect my 4 repeaters to forward all my requests within +-400ms. If my understanding of this change is correct, it means that I will only have received all 4 the requests in 20min!

As per our discussion, the limiting factor here is the fact that Repeaters are checked every 5 min. Once one fails (all others except the first one in this case), it is kicked down the road to the next 5 min slot, irrespective of its interval time. One option is to have a celery task dedicated to handling repeaters with throttled connections, have them scheduled with smaller periods so that if there are multiple requests that needs to go through, they are not kicked down the road into the far future the whole time, but rather is allowed to go through relatively quick after each other.

I wonder: another (not so thought through) way is to maybe use the asyncio (or similar) library to allow these pending requests to be "waited out" without waiting a whole 5 mins. I don't know..it's late

Reply:
A two pronged approach is probably best here:
* reduce the time between retries via the queue (consult with SaaS but I don't see why we couldn't try reduce the time and monitor with DD)
* build in some local retry with backoff into the sending task (potentially only activate this for short delay times - or you could use `connection_settings.get_next_attempt` to determine if local retry should happen).

Reply:
On challenge with this kind of throttling is that, because there is no coordination or prioritization, it's possible for one record to get stuck and never get a slot to send because other records get there first.

It's also possible that multiple records could fire concurrently if they all get past the first check before the throttle time is set.

Reply:
> On challenge with this kind of throttling is that, because there is no coordination or prioritization, it's possible for one record to get stuck and never get a slot to send because other records get there first.

Exactly. How the rescheduling is handled basically determines the success of this feature. Get that wrong and this throttling makes it worse.

Thanks for the suggestions; I'll let the them simmer in my mind so long.